Kuadrant/authorino-operator
# Support for Authorino's Raw HTTP Authorization interface

- Adds new field `spec.listener.ports{grpc:int,http:int}` to the API
- Deprecated `spec.listener.port`
- Reconciles the state of the fields `spec.listener.ports` (and now deprecated `spec.listener.port`) and `spec.oidcServer.port` updating the corresponding service specs and env vars injected in the deployment
- Bugfix: ensure services are created with the chosen port numbers instead of always the default ones

Closes #58

### Verification steps

Deploy this branch:

```sh
kind create cluster --name authorino
make docker-build OPERATOR_IMAGE=authorino-operator:local
kind load docker-image authorino-operator:local --name authorino
kubectl create namespace authorino-operator
make install deploy OPERATOR_IMAGE=authorino-operator:local
```

After applying each version of the Authorino CR below, verify if the services and env vars were reconcilled properly:

```sh
# default port numbers
kubectl apply -f -<<EOF
apiVersion: operator.authorino.kuadrant.io/v1beta1
kind: Authorino
metadata:
  name: authorino
spec:
  listener:
    tls:
      enabled: false
  oidcServer:
    tls:
      enabled: false
EOF
```

```sh
# custom raw http authz port number
kubectl apply -f -<<EOF
apiVersion: operator.authorino.kuadrant.io/v1beta1
kind: Authorino
metadata:
  name: authorino
spec:
  listener:
    ports:
      http: 5000
    tls:
      enabled: false
  oidcServer:
    tls:
      enabled: false
EOF
```

```sh
# raw http authz disabled
kubectl apply -f -<<EOF
apiVersion: operator.authorino.kuadrant.io/v1beta1
kind: Authorino
metadata:
  name: authorino
spec:
  listener:
    ports:
      http: 0
    tls:
      enabled: false
  oidcServer:
    tls:
      enabled: false
EOF
```

```sh
# custom grpc auth port number and raw http authz disabled
kubectl apply -f -<<EOF
apiVersion: operator.authorino.kuadrant.io/v1beta1
kind: Authorino
metadata:
  name: authorino
spec:
  listener:
    ports:
      grpc: 50052
      http: 0
    tls:
      enabled: false
  oidcServer:
    tls:
      enabled: false
EOF
```

```sh
# custom grpc auth port number (using deprecated api field) and raw http authz disabled
kubectl apply -f -<<EOF
apiVersion: operator.authorino.kuadrant.io/v1beta1
kind: Authorino
metadata:
  name: authorino
spec:
  listener:
    port: 30001
    ports:
      http: 0
    tls:
      enabled: false
  oidcServer:
    tls:
      enabled: false
EOF
```

```sh
# custom oidc port number
kubectl apply -f -<<EOF
apiVersion: operator.authorino.kuadrant.io/v1beta1
kind: Authorino
metadata:
  name: authorino
spec:
  listener:
    port: 30001
    ports:
      http: 0
    tls:
      enabled: false
  oidcServer:
    port: 9000
    tls:
      enabled: false
EOF
```

```sh
# grpc disabled, custom raw http authz port number, and custom oidc port number
kubectl apply -f -<<EOF
apiVersion: operator.authorino.kuadrant.io/v1beta1
kind: Authorino
metadata:
  name: authorino
spec:
  listener:
    ports:
      grpc: 0
      http: 3000
    tls:
      enabled: false
  oidcServer:
    port: 9000
    tls:
      enabled: false
EOF
```

Try the service with the latest port settings:

```sh
kubectl apply -f -<<EOF
apiVersion: v1
kind: Secret
metadata:
  name: wristband-signing-key
stringData:
  key.pem: |
    -----BEGIN EC PRIVATE KEY-----
    MHcCAQEEIDHvuf81gVlWGo0hmXGTAnA/HVxGuH8vOc7/8jewcVvqoAoGCCqGSM49
    AwEHoUQDQgAETJf5NLVKplSYp95TOfhVPqvxvEibRyjrUZwwtpDuQZxJKDysoGwn
    cnUvHIu23SgW+Ee9lxSmZGhO4eTdQeKxMA==
    -----END EC PRIVATE KEY-----
type: Opaque
---
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: wristband
spec:
  hosts:
  - wristband.127.0.0.1.nip.io
  response:
  - name: wristband
    wristband:
      customClaims:
      - name: roles
        value:
        - admin
      issuer: http://authorino-authorino-oidc.default.svc.cluster.local:9000/default/wristband/wristband
      signingKeyRefs:
      - algorithm: ES256
        name: wristband-signing-key
EOF
```

```sh
kubectl port-forward services/authorino-authorino-oidc 9000:9000 &;PID=$!

curl http://localhost:9000/default/wristband/wristband/.well-known/openid-connect/certs
# HTTP/1.1 200 OK

kill $PID
```

```sh
kubectl port-forward services/authorino-authorino-authorization 3000:3000 &;PID=$!

curl -H 'Content-type: application/json' -d '{}' http://wristband.127.0.0.1.nip.io:3000/check -i
# HTTP/1.1 200 OK

kill $PID
```
